projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+
+#include "imgproc/opencv_unwrap_360.h"
+#include "video/panoramic.h"
+
+#include <sys/stat.h>
+#include <ctime>
+
+
+
+
+
+
+// Standard C++ includes
+#include <iostream>
+
+using namespace BoBRobotics;
+using namespace BoBRobotics::ImgProc;
+using namespace BoBRobotics::Video;
+
+
+#define WIDTH 192
+#define HEIGHT 72
+
+#define IS_UNWRAP true
+#define UPDATE_INTERVAL 200
+
+
+void readCameraThreadFunc(BoBRobotics::Video::Input *cam,
+                        cv::Mat &img,
+                        std::atomic<bool> &shouldQuit,
+                        std::mutex &mutex)
+{
+
+    //auto unwrapper = cam->createUnwrapper(cv::Size(WIDTH,HEIGHT));
+
+    while(!shouldQuit) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(5));
+        std::lock_guard<std::mutex> lock(mutex);
+        cam->readFrame(img);
+
+    }
+
+
+}
+
+void saveThread(cv::Mat img, std::string fileName) {
+    cv::imwrite(fileName, img);
+}
+
+int bobMain(int argc, char* argv[])
+{
+
+    using degree_t = units::angle::degree_t;
+    std::vector<cv::Mat> images;
+    std::vector<std::string> fileNames;
+
+    // setting up gps
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    BoBRobotics::GPS::Gps gps;
+    gps.connect(path_linux);
+    BoBRobotics::BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+    std::atomic<bool> shouldQuit{false};
+    std::mutex mex;
+
+
+    time_t rawtime;
+    struct tm * timeinfo;
+    char buffer[80];
+
+    time (&rawtime);
+    timeinfo = localtime(&rawtime);
+
+    strftime(buffer,sizeof(buffer),"%d-%m-%Y %H:%M:%S",timeinfo);
+    std::string str(buffer);
+
+    int seconds_to_run = 30; // default value
+    if (argc > 1) {
+        seconds_to_run = stoi(std::string(argv[1]));
+        std::cout << "running for " << seconds_to_run << " seconds" << std::endl;
+
+    }
+
+
+
+    // check to see if we get valid coordinates by checking GPS quality
+    while (numTrials > 0) {
+        try {
+            if (gps.getGPSData().gpsQuality != BoBRobotics::GPS::GPSQuality::INVALID) {
+                std::cout << " we have a valid measurement" << std::endl;
+                break;
+            }
+        } catch(...) {
+            std::cout << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]" << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+    if (numTrials == 0) {
+        std::cout << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program " << std::endl;
+        exit(1);
+    }
+
+    auto gps_time =  gps.getGPSData().time;
+    int gps_hour = gps_time.hour;
+    int gps_minute = gps_time.minute;
+    int gps_second = gps_time.second;
+
+    std::string folderName;
+    //std::cin >> folderName;
+
+    folderName = str;
+
+    const char *c  = folderName.c_str();
+    mkdir(c, 0777);
```

Comment:
I'd check the return code for this to make sure the directory has actually been created, e.g.:
```c++
BOB_ASSERT(mkdir(c, 0777) == 0);
```

I'm guessing that you're making this world-writable because you're running your program as root?

Reply:
you should _*never*_ have to run stuff like this as root! For serial stuff, you need to add the user to dialout and for I2C, you need to follow these instructions https://forums.developer.nvidia.com/t/accessing-gpio-without-using-sudo-su/38774

Reply:
Yeah, it's a bit of a hack. Adding a udev rule is definitely the way to go.